reporter/reporter.go

```diff
@@ -0,0 +1,237 @@
+	ps := h.Percentiles(r.percentiles)
+	r.sender.SendMetric(r.prepareName(name+".count"), float64(h.Count()), 0, r.source, tags)
+	r.sender.SendMetric(r.prepareName(name+".min"), float64(h.Min()), 0, r.source, tags)
+	r.sender.SendMetric(r.prepareName(name+".max"), float64(h.Max()), 0, r.source, tags)
+	r.sender.SendMetric(r.prepareName(name+".mean"), h.Mean(), 0, r.source, tags)
+	r.sender.SendMetric(r.prepareName(name+".std-dev"), h.StdDev(), 0, r.source, tags)
+	for psIdx, psKey := range r.percentiles {
+		key := strings.Replace(strconv.FormatFloat(psKey*100.0, 'f', -1, 64), ".", "", 1)
+		r.sender.SendMetric(r.prepareName(name+"."+key+"-percentile"), ps[psIdx], 0, r.source, tags)
+	}
+}
+
+func (r *reporter) reportMeter(name string, metric metrics.Meter, tags map[string]string) {
+	m := metric.Snapshot()
+	r.sender.SendMetric(r.prepareName(name+".count"), float64(m.Count()), 0, r.source, tags)
+	r.sender.SendMetric(r.prepareName(name+".one-minute"), m.Rate1(), 0, r.source, tags)
+	r.sender.SendMetric(r.prepareName(name+".five-minute"), m.Rate5(), 0, r.source, tags)
+	r.sender.SendMetric(r.prepareName(name+".fifteen-minute"), m.Rate15(), 0, r.source, tags)
+	r.sender.SendMetric(r.prepareName(name+".mean"), m.RateMean(), 0, r.source, tags)
+}
+
+func (r *reporter) reportTimer(name string, metric metrics.Timer, tags map[string]string) {
+	t := metric.Snapshot()
+	du := float64(r.durationUnit)
+	ps := t.Percentiles(r.percentiles)
+	r.sender.SendMetric(r.prepareName(name+".count"), float64(t.Count()), 0, r.source, tags)
+	r.sender.SendMetric(r.prepareName(name+".min"), float64(t.Min()/int64(du)), 0, r.source, tags)
+	r.sender.SendMetric(r.prepareName(name+".max"), float64(t.Max()/int64(du)), 0, r.source, tags)
+	r.sender.SendMetric(r.prepareName(name+".mean"), t.Mean()/du, 0, r.source, tags)
+	r.sender.SendMetric(r.prepareName(name+".std-dev"), t.StdDev()/du, 0, r.source, tags)
+	for psIdx, psKey := range r.percentiles {
+		key := strings.Replace(strconv.FormatFloat(psKey*100.0, 'f', -1, 64), ".", "", 1)
+		r.sender.SendMetric(r.prepareName(name+"."+key+"-percentile"), ps[psIdx]/du, 0, r.source, tags)
+	}
+	r.sender.SendMetric(r.prepareName(name+".one-minute"), t.Rate1(), 0, r.source, tags)
+	r.sender.SendMetric(r.prepareName(name+".five-minute"), t.Rate5(), 0, r.source, tags)
+	r.sender.SendMetric(r.prepareName(name+".fifteen-minute"), t.Rate15(), 0, r.source, tags)
+	r.sender.SendMetric(r.prepareName(name+".mean-rate"), t.RateMean(), 0, r.source, tags)
+}
+
+func (r *reporter) prepareName(name string, suffix ...string) string {
+	if len(r.prefix) > 0 {
+		name = r.prefix + "." + name
+	}
+
+	if r.addSuffix {
+		for _, s := range suffix {
+			name += "." + s
+		}
+	}
+
+	return name
+}
+
+func (r *reporter) Start() {
+	if r.ticker == nil {
+		r.ticker = time.NewTicker(r.interval)
+		go func() {
+			for range r.ticker.C {
+				r.report()
+			}
+		}()
+	}
+}
+
+func (r *reporter) Stop() {
+	if r.ticker != nil {
+		r.ticker.Stop()
```

Comment:
This only stops the clock. Use a channel to signal done to the go routine, and select in the goroutine between the ticker and the channel. Current implementation will leak go routines.